Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/TicketService.hs

```diff
@@ -594,3 +668,254 @@ postTicketBookingsUpdateSeats _ TicketBookingUpdateSeatsReq {..} = do
+    findTicketBookingServiceCategoryToCancel ticketServiceCategoryId ticketBookingService = do
+      currentTime <- getCurrentTime
+      mbTBServiceCategory <- QTBSC.findByTBServiceAndServiceCategory ticketBookingService.id (Just ticketServiceCategoryId.getId) Nothing
+      case mbTBServiceCategory of
+        Just tbServiceCategory -> do
+          ticketBooking <- QTB.findById ticketBookingService.ticketBookingId >>= fromMaybeM (TicketBookingNotFound ticketBookingService.ticketBookingId.getId)
+          person <- runInReplica $ QP.findById ticketBooking.personId >>= fromMaybeM (PersonNotFound ticketBooking.personId.getId)
+          if isValidateTicketBooking ticketBooking ticketBookingService currentTime
+            then pure $ Just (person, ticketBooking, ticketBookingService, tbServiceCategory)
+            else pure Nothing
+        Nothing -> pure Nothing
+      where
+        isValidateTicketBooking :: DTTB.TicketBooking -> DTB.TicketBookingService -> UTCTime -> Bool
+        isValidateTicketBooking ticketBooking ticketBookingService' currentTime
+          | ticketBookingService'.status /= DTB.Confirmed = False
+          | ticketBooking.status /= DTTB.Booked = False
+          | maybe False (currentTime >) ticketBookingService'.expiryDate = False
+          | otherwise = True
+
+    cancelAndRefundConfirmedBookingService :: (Domain.Types.Person.Person, Domain.Types.TicketBooking.TicketBooking, DTB.TicketBookingService, DTB.TicketBookingServiceCategory) -> Environment.Flow ()
+    cancelAndRefundConfirmedBookingService (person, ticketBooking, ticketBookingService, tbServiceCategory) = do
+      Redis.whenWithLockRedis (ticketBookingCancellationLockKey ticketBooking.id.getId) 60 $ do
+        let refundAmount = tbServiceCategory.amount.amount - fromMaybe (HighPrecMoney 0) tbServiceCategory.amountToRefund
+        _ <- intializeRefundProcess ticketBooking refundAmount person.merchantId person.merchantOperatingCityId
+        let cancelledSeats = Just $ tbServiceCategory.bookedSeats - fromMaybe 0 tbServiceCategory.cancelledSeats
+        QTBSC.updateCancellationDetailsById cancelledSeats (Just tbServiceCategory.amount.amount) (Just DTB.Merchant) tbServiceCategory.id
+
+        peopleBookingCategories <- QTBPC.findAllByServiceCategoryId tbServiceCategory.id
+        mapM_ updatePeopleCancelledUnitsAndAmount peopleBookingCategories
+        sendSms person ticketBookingService.id.getId
+        Notifications.notifyTicketCancelled ticketBooking person
+      where
+        updatePeopleCancelledUnitsAndAmount :: DTB.TicketBookingPeopleCategory -> Environment.Flow ()
+        updatePeopleCancelledUnitsAndAmount pc = do
+          let refundAmountForPeopleCategory = Just $ fromIntegral (pc.numberOfUnits - fromMaybe 0 pc.numberOfUnitsCancelled) * pc.pricePerUnit.amount + fromMaybe (HighPrecMoney 0) pc.amountToRefund
```

Comment:
it is for keeping track of the refund amount for people category, suppose at x1 time 3 out of 10 people in a book is cancelled it will have some x1 charges based on x1 time and again x2 time again 4 people out of remaining 7 cancel it will have x2 charges based on x2 time 
so to keep track we need to add field there